// contentful-batch-libs/dist/push/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pushToSpace = exports.getEntityName = exports.publishing = exports.deletion = exports.creation = exports.assets = undefined;

var _getEntityName = require('./get-entity-name');

Object.defineProperty(exports, 'getEntityName', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_getEntityName).default;
  }
});

var _pushToSpace = require('./push-to-space');

Object.defineProperty(exports, 'pushToSpace', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_pushToSpace).default;
  }
});

var _assets = require('./assets');

var assets = _interopRequireWildcard(_assets);

var _creation = require('./creation');

var creation = _interopRequireWildcard(_creation);

var _deletion = require('./deletion');

var deletion = _interopRequireWildcard(_deletion);

var _publishing = require('./publishing');

var publishing = _interopRequireWildcard(_publishing);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.assets = assets;
exports.creation = creation;
exports.deletion = deletion;
exports.publishing = publishing;
